jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java

```diff
@@ -0,0 +1,540 @@
+        if (cachingHttpContent != null)
+        {
+            cachingHttpContent.setLastAccessedNanos(NanoTime.now());
+            if (cachingHttpContent.isValid())
+                return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+            else
+                removeFromCache(cachingHttpContent);
+        }
+
+        HttpContent httpContent = _authority.getContent(path);
+        if (isCacheable(httpContent))
+        {
+            AtomicBoolean wasAdded = new AtomicBoolean(false);
+            cachingHttpContent = _cache.computeIfAbsent(path, p ->
+            {
+                wasAdded.set(true);
+                CachingHttpContent cachingContent = (httpContent == null) ? newNotFoundContent(p) : newCachedContent(p, httpContent);
+                _cachedSize.addAndGet(cachingContent.getContentLengthValue());
+                return cachingContent;
+            });
+
+            if (wasAdded.get())
+                shrinkCache();
+            return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+        }
+        return httpContent;
+    }
+
+    protected CachingHttpContent newCachedContent(String p, HttpContent httpContent)
+    {
+        return new CachedHttpContent(p, httpContent);
+    }
+
+    protected CachingHttpContent newNotFoundContent(String p)
+    {
+        return new NotFoundHttpContent(p);
+    }
+
+    protected interface CachingHttpContent extends HttpContent
+    {
+        long getLastAccessedNanos();
+
+        void setLastAccessedNanos(long nanosTime);
+
+        String getKey();
+
+        boolean isValid();
+    }
+
+    protected class CachedHttpContent extends HttpContentWrapper implements CachingHttpContent
+    {
+        private final RetainableByteBuffer _buffer;
+        private final String _cacheKey;
+        private final HttpField _etagField;
+        private final long _contentLengthValue;
+        private final AtomicLong _lastAccessed = new AtomicLong();
+        private final Set<CompressedContentFormat> _compressedFormats;
+        private final String _lastModifiedValue;
+        private final String _characterEncoding;
+        private final MimeTypes.Type _mimeType;
+        private final HttpField _contentLength;
+        private final Instant _lastModifiedInstant;
+        private final HttpField _lastModified;
+
+        public CachedHttpContent(String key, HttpContent httpContent)
+        {
+            super(httpContent);
+            _cacheKey = key;
+
+            // TODO: do all the following lazily and asynchronously.
+            HttpField etagField = httpContent.getETag();
+            String eTagValue = httpContent.getETagValue();
+            if (StringUtil.isNotBlank(eTagValue))
+            {
+                etagField = new PreEncodedHttpField(HttpHeader.ETAG, eTagValue);
+            }
+            _etagField = etagField;
+
+            // Map the content into memory if possible.
```

Comment:
```suggestion
            // Read the content into memory if possible.
```

What if the buffer was a `FileMappedHttpContentFactory.FileMappedContent`, wont you also read it?

Reply:
Yes, but I don't think we should be reading it. The check in `CachingContentFactory.isCacheable`
```java
if (httpContent instanceof FileMappedHttpContentFactory.FileMappedContent)
             return true;
```
indicates that we don't count these towards the size so we shouldn't be copying it into indirect buffer.

I will change this to 
```java
// Read the content into memory if the HttpContent does not already have a buffer.
RetainableByteBuffer buffer = httpContent.getBuffer();
if (buffer != null)
{
    ...
```